test/test_type_promotion.py

```diff
@@ -437,6 +438,126 @@ def test_transpose(self, device):
+            self.assertRaises(RuntimeError, lambda: op(dense1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, dense2))
+            return
+
+        expected = op(dense1.clone(), dense2)
+        precision = self._get_precision(expected.dtype, coalesced)
+        test_tensors = [expected, dense1, sparse1, dense2, sparse2]
+        e, d1, s1, d2, s2 = [x.clone() for x in test_tensors] if inplace else test_tensors
+
+        # Test op(sparse, sparse)
+        if op_name != 'div':
+            sparse = op(s1, s2)
+            self.assertEqual(sparse.dtype, e.dtype)
+            self.assertEqual(e, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            self.assertRaises(RuntimeError, lambda: op(s1, s2).to_dense())
+
+        # Test op(dense, sparse)
+        if add_sub:
+            if inplace:
+                e, d1, s1, d2, s2 = [x.clone() for x in test_tensors]
+            dense_sparse = op(d1, s2)
+            self.assertEqual(e, dense_sparse, prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            # mul: Didn't find kernel to dispatch to for operator 'aten::_nnz'
+            self.assertRaises(RuntimeError, lambda: op(d1, s2))
+
+        # Test op(sparse, dense) not supported for any ops:
+        # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+        # sparse division only supports division by a scalar
+        # mul: Didn't find kernel to dispatch to for operator 'aten::_nnz'.
+        self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        # Test op(sparse, scalar)
+        if not add_sub and not (self.device_type == 'cpu' and dtype1 == torch.half):
+            if inplace:
+                e, d1, s1, d2, s2 = [x.clone() for x in test_tensors]
+            scalar = d2.view(d2.numel())[0].item()
+
+            sparse = op(s1, scalar)
+            dense_scalar = op(d1, scalar)
+            self.assertEqual(sparse.dtype, dense_scalar.dtype)
+            self.assertEqual(dense_scalar, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+            # "mul_cpu" / "div_cpu" not implemented for 'Half'
+            self.assertRaises(RuntimeError, lambda: op(s1, d2.view(d2.numel())[0].item()))
+
+    def _run_all_tests_for_sparse_op(self, op_name, device):
+        dtypes = torch.testing.get_all_math_dtypes(device)
+        for dtype1, dtype2 in itertools.product(dtypes, dtypes):
+            for inplace, coalesced in itertools.product([True, False], [True, False]):
+                self._test_sparse_op(op_name, inplace, dtype1, dtype2, device, coalesced)
+
+    @onlyOnCPUAndCUDA
+    def test_sparse_add(self, device):
```

Comment:
Out of curiosity, how long do these take to run now that we've changed the sizes to [5, 5, 5]?

Reply:
I don't remember what the old results were but I don't think it made a big difference. TestTypePromotionCPU.test_sparse_add_cpu takes 6 seconds and the cuda variant takes 30. I think the cuda variant takes so long due to copying small tensors to CUDA, and we can probably improve on this by refactoring test tensor generation in https://github.com/pytorch/pytorch/issues/30611.

Reply:
Thank you for marking these to only run on CPU and CUDA!